File: triatlon-java/src/main/java/ro/ubbcluj/scs/dfar2166/triatlonjava/service/TriatlonService.java
package ro.ubbcluj.scs.dfar2166.triatlonjava.service;



import ro.ubbcluj.scs.dfar2166.triatlonjava.model.Result;
import ro.ubbcluj.scs.dfar2166.triatlonjava.model.activity.RaceType;
import ro.ubbcluj.scs.dfar2166.triatlonjava.model.activity.AthletePoints;
import ro.ubbcluj.scs.dfar2166.triatlonjava.model.person.Athlete;
import ro.ubbcluj.scs.dfar2166.triatlonjava.model.person.Referee;
import ro.ubbcluj.scs.dfar2166.triatlonjava.repository.AthleteRepository;
import ro.ubbcluj.scs.dfar2166.triatlonjava.repository.RaceRepository;
import ro.ubbcluj.scs.dfar2166.triatlonjava.repository.RefereeRepository;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

// TriatlonService

public class TriatlonService {

    // Private Properties

    private final RaceRepository raceRepository;
    private final RefereeRepository refereeRepository;
    private final AthleteRepository athleteRepository;
    private Referee loggedInReferee = null;

    // Lifecycle

    public TriatlonService(RaceRepository raceRepository, RefereeRepository refereeRepository, AthleteRepository athleteRepository) {
        this.raceRepository = raceRepository;
        this.refereeRepository = refereeRepository;
        this.athleteRepository = athleteRepository;
    }

    // Instance Methods

    public boolean logInReferee(String email, String password) {
        Referee referee = this.refereeRepository.findByEmail(email);
        if (referee == null) return false;
        var isLoggedIn = referee.getEmail().equals(email) && referee.getPassword().equals(password);
        if (isLoggedIn) {
            loggedInReferee = referee;
            this.raceRepository.setRaceType(referee.getRaceType());
        }
        return isLoggedIn;
    }

    public void logOutReferee() {
        loggedInReferee = null;
        this.raceRepository.setRaceType(null);
    }

    public Referee getCurrentReferee() {
        return loggedInReferee;
    }

    public List<Athlete> getAthletes() {
        List<Athlete> athletes = new ArrayList<>();
        this.athleteRepository.findAll().forEach(athletes::add);
        return athletes;
    }

    public List<Result> getAthletesWithTotalPoints() {
        List<Result> athletesWithResults = new ArrayList<>();
        this.athleteRepository.findAll().forEach(athlete -> {
            var points = 0;
            for (var value : RaceType.values()) {
                raceRepository.setRaceType(value);
                AthletePoints athletePoints = raceRepository.findByAthleteId(athlete.getId());
                if (athletePoints != null) points += athletePoints.getPoints();
            }
            Result athleteResult = new Result(athlete.getId(), athlete.getFullName(), athlete.getFullNameReversed(), points);
            athletesWithResults.add(athleteResult);
        });
        raceRepository.setRaceType(loggedInReferee.getRaceType());
        return athletesWithResults.stream().sorted(Comparator.comparing(Result::getNameReversed)).collect(Collectors.toList());
    }

    public void addResult(Integer athleteId, Integer points) {
        AthletePoints athletePoints = this.raceRepository.findByAthleteId(athleteId);
        if (athletePoints == null) {
            athletePoints = new AthletePoints(athleteId, points);
            this.raceRepository.save(athletePoints);
        } else {
            athletePoints.setPoints(points);
            this.raceRepository.update(athletePoints);
        }
    }

    public List<Result> getParticipantsWithResultInRace() {
        List<Result> results = new ArrayList<>();
        this.raceRepository.findAllWithPoints().forEach(athletePoints -> {
            Athlete athlete = this.athleteRepository.findOne(athletePoints.getAthleteId());
            Result result = new Result(athlete.getId(), athlete.getFullName(), athlete.getFullNameReversed(), athletePoints.getPoints());
            results.add(result);
        });
        return results.stream().sorted(Comparator.comparing(Result::getPoints).reversed()).collect(Collectors.toList());
    }

}

File: triatlon-cs/Service/TriatlonService.cs

using System;
using System.Collections.Generic;
using System.Linq;

public class TriatlonService
{
    
    // Properties

    private RaceRepository _raceRepository;
    private IRefereeRepository _refereeRepository;
    private IAthleteRepository _athleteRepository;
    private Referee _loggedInReferee = null;
    
    // Lifecycle

    public TriatlonService(RaceRepository raceRepository, IRefereeRepository refereeRepository,
        IAthleteRepository athleteRepository)
    {
        _raceRepository = raceRepository;
        _refereeRepository = refereeRepository;
        _athleteRepository = athleteRepository;
    }
    
    // Instance Methods

    public bool LogInReferee(string email, string password)
    {
        Referee referee = _refereeRepository.FindByEmail(email);
        if (referee == null) return false;
        var isLoggedIn = referee.Email.Equals(email) && referee.Password.Equals(password);
        if (isLoggedIn)
        {
            _loggedInReferee = referee;
            _raceRepository.RaceType = referee.RaceType;
        }
        return isLoggedIn;
    }

    public void LogOutReferee()
    {
        _loggedInReferee = null;
    }

    public Referee CurrentReferee
    {
        get => _loggedInReferee;
    }

    public List<Athlete> GetAthletes()
    {
        List<Athlete> athletes = new List<Athlete>();
        foreach (var athlete in _athleteRepository.FindAll())
        {
            athletes.Add(athlete);
        }
        return athletes;
    }

    public List<Result> GetAthletesWithTotalPoints()
    {
        List<Result> athletesWithResults = new List<Result>();
        foreach (var athlete in _athleteRepository.FindAll())
        {
            var points = 0;
            foreach (var value in Enum.GetValues(typeof(RaceType)))
            {
                _raceRepository.RaceType = (RaceType) value;
                AthletePoints athletePoints = _raceRepository.FindByAthleteId(athlete.Id);
                if (athletePoints != null) points += athletePoints.Points;
            }

            Result athleteResult = new Result(athlete.Id, athlete.FullName, athlete.FullNameReversed, points);
            athletesWithResults.Add(athleteResult);
        }
        _raceRepository.RaceType = _loggedInReferee.RaceType;
        return athletesWithResults.OrderBy(r=> r.ReversedName).ToList();
    }

    public void AddResult(int athleteId, int points)
    {
        AthletePoints athletePoints = _raceRepository.FindByAthleteId(athleteId);
        if (athletePoints == null)
        {
            athletePoints = new AthletePoints(athleteId, points);
            _raceRepository.Save(athletePoints);
        }
        else
        {
            athletePoints.Points = points;
            _raceRepository.Update(athletePoints);
        }
    }

    public List<Result> GetParticipantsWithResultInRace()
    {
        List<Result> results = new List<Result>();
        foreach (var athletePoints in _raceRepository.FindAllWithPoints())
        {
            Athlete athlete = _athleteRepository.FindOne(athletePoints.AthleteId);
            Result result = new Result(athlete.Id, athlete.FullName, athlete.FullNameReversed, athletePoints.Points);
            results.Add(result);
        }
        return results.OrderByDescending(r=> r.Points).ToList();
    }

}